crypto/secp256k1.re

```diff
@@ -0,0 +1,123 @@
+  let rand_value =
+    Mirage_crypto_rng.generate(32) |> Cstruct.to_bytes |> Bigstring.of_bytes;
+  let randomized = Context.randomize(c, rand_value);
+  randomized
+    ? c : failwith("Secp256k1 context randomization failed. Aborting.");
+};
+
+module Key = {
+  type t = Libsecp256k1.Key.t(Libsecp256k1.Key.public);
+  let equal = (a, b) => {
+    Libsecp256k1.Key.equal(a, b);
+  };
+  let compare = (a, b) => {
+    Bigstring.compare(
+      Libsecp256k1.Key.buffer(a),
+      Libsecp256k1.Key.buffer(b),
+    );
+  };
+  let size = Libsecp256k1.Key.compressed_pk_bytes;
+  let prefix = Base58.Prefix.secp256k1_public_key;
+  let encoding = {
+    let to_bytes = t =>
+      Libsecp256k1.Key.to_bytes(context, t) |> Bigstring.to_bytes;
+    let of_bytes_exn = b =>
+      Libsecp256k1.Key.read_pk_exn(context, Bigstring.of_bytes(b));
+    Data_encoding.(conv(to_bytes, of_bytes_exn, Fixed.bytes(size)));
+  };
+  let of_secret = sk => Libsecp256k1.Key.neuterize_exn(context, sk);
+  let to_raw = t => Key.to_bytes(context, t) |> Bigstring.to_string;
+
+  let of_raw = string =>
+    Key.read_pk(context, string |> Bigstring.of_string) |> Result.to_option;
+
+  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+  let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+};
+module Key_hash = {
+  [@deriving (ord, eq)]
+  type t = BLAKE2B_20.t;
+
+  let hash_key = t => BLAKE2B_20.hash(Key.to_raw(t));
+
+  let encoding = {
+    let name = "Secp256k1.Public_key_hash";
+    Data_encoding.(obj1(req(name, blake2b_20_encoding)));
+  };
+  let prefix = Base58.Prefix.secp256k1_public_key_hash;
+  let to_raw = BLAKE2B_20.to_raw_string;
+  let of_raw = BLAKE2B_20.of_raw_string;
+  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+  let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+};
+module Secret = {
+  type t = Libsecp256k1.Key.t(Libsecp256k1.Key.secret);
+
+  let equal = (a, b) => {
+    Libsecp256k1.Key.equal(a, b);
+  };
+  let compare = (a, b) => {
+    Bigstring.compare(
+      Libsecp256k1.Key.buffer(a),
+      Libsecp256k1.Key.buffer(b),
+    );
+  };
+  let _size = Libsecp256k1.Key.secret_bytes;
+  let prefix = Base58.Prefix.secp256k1_secret_key;
+  let to_raw = t =>
+    Libsecp256k1.Key.to_bytes(context, t) |> Bigstring.to_string;
+  let of_raw = string => {
+    Libsecp256k1.Key.read_sk(context, string |> Bigstring.of_string)
+    |> Result.to_option;
+  };
+
+  let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+  let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+};
+module Signature = {
+  type t = Sign.t(Sign.plain);
+
+  let equal = (a, b) => {
+    Sign.equal(a, b);
+  };
+  let compare = (a, b) => {
+    Bigstring.compare(Sign.buffer(a), Sign.buffer(b));
+  };
+  let _size = Sign.plain_bytes;
+  let prefix = Base58.Prefix.secp256k1_signature;
+  let to_raw = t =>
+    Sign.to_bytes(~der=false, context, t) |> Bigstring.to_string;
```

Comment:
what is `der`?

Reply:
Encoding rules IIUC
https://datatracker.ietf.org/doc/html/rfc5915#ref-X.690